src/MiddlewareStack.php

```diff
@@ -0,0 +1,55 @@
+<?php
+
+namespace React\Http;
+
+use Psr\Http\Message\ResponseInterface;
+use Psr\Http\Message\ServerRequestInterface;
+use React\Promise;
+use React\Promise\PromiseInterface;
+
+final class MiddlewareStack implements MiddlewareStackInterface
+{
+    /**
+     * @var ResponseInterface
+     */
+    private $defaultResponse;
+
+    /**
+     * @var MiddlewareInterface[]
+     */
+    private $middlewares = array();
+
+    /**
+     * @param ResponseInterface $response
+     * @param MiddlewareInterface[] $middlewares
+     */
+    public function __construct(ResponseInterface $response, array $middlewares)
+    {
+        $this->defaultResponse = $response;
+        $this->middlewares = $middlewares;
+    }
+
+    /**
+     * @param ServerRequestInterface $request
+     * @return PromiseInterface<ResponseInterface>
+     */
+    public function process(ServerRequestInterface $request)
+    {
+        if (count($this->middlewares) === 0) {
+            return Promise\resolve($this->defaultResponse);
+        }
+
+        $middlewares = $this->middlewares;
+        $middleware = array_shift($middlewares);
+
+        return Promise\resolve(
+            $middleware->process(
+                $request,
+                new self(
+                    $this->defaultResponse,
+                    $middlewares
+                )
+            )
+        );
+    }
```

Comment:
To prevent exceptions thrown from `process()` bubbling out from here, this should be wrapped like this to turn them into rejections:

```php
return new Promise\Promise(function ($resolve, $reject) use ($request, $middleware, $middleware) {
    $response = $middleware->process(
        $request,
        new self(
            $this->defaultResponse,
            $middlewares
        )
    );

    $resolve($response);
});
```